src/main/java/net/fabricmc/loom/configuration/ifaceinject/InterfaceInjectionProcessor.java

```diff
@@ -0,0 +1,273 @@
+			}
+		};
+	}
+
+	@Override
+	public boolean isInvalid(File file) {
+		byte[] hash = ZipUtil.unpackEntry(file, HASH_FILENAME);
+
+		if (hash == null) {
+			return true;
+		}
+
+		return !Arrays.equals(inputHash, hash);
+	}
+
+	private List<InjectedInterface> getInjectedInterfaces() {
+		List<InjectedInterface> result = new ArrayList<>();
+
+		for (RemappedConfigurationEntry entry : Constants.MOD_COMPILE_ENTRIES) {
+			// Only apply injected interfaces from mods that are part of the compile classpath
+			if (!entry.compileClasspath()) {
+				continue;
+			}
+
+			Set<File> artifacts = extension.getLazyConfigurationProvider(entry.sourceConfiguration())
+					.get()
+					.resolve();
+
+			for (File artifact : artifacts) {
+				result.addAll(InjectedInterface.fromModJar(artifact.toPath()));
+			}
+		}
+
+		return result;
+	}
+
+	private record InjectedInterface(String modId, String className, String ifaceName) {
+		/**
+		 * Reads the injected interfaces contained in a mod jar, or returns null if there is none.
+		 */
+		public static List<InjectedInterface> fromModJar(Path modJarPath) {
+			byte[] modJsonBytes = ZipUtil.unpackEntry(modJarPath.toFile(), "fabric.mod.json");
+
+			if (modJsonBytes == null) {
+				return Collections.emptyList();
+			}
+
+			JsonObject jsonObject = new Gson().fromJson(new String(modJsonBytes, StandardCharsets.UTF_8), JsonObject.class);
+
+			String modId = jsonObject.get("id").getAsString();
+
+			if (!jsonObject.has("custom")) {
+				return Collections.emptyList();
+			}
+
+			JsonObject custom = jsonObject.getAsJsonObject("custom");
+
+			if (!custom.has("loom:injected_interfaces")) {
+				return Collections.emptyList();
+			}
+
+			JsonObject addedIfaces = custom.getAsJsonObject("loom:injected_interfaces");
+
+			List<InjectedInterface> result = new ArrayList<>();
+
+			for (String className : addedIfaces.keySet()) {
+				JsonArray ifaceNames = addedIfaces.getAsJsonArray(className);
+
+				for (JsonElement ifaceName : ifaceNames) {
+					result.add(new InjectedInterface(modId, className, ifaceName.getAsString()));
+				}
+			}
+
+			return result;
+		}
+	}
+
+	private static class InjectingClassVisitor extends ClassVisitor {
+		private final List<InjectedInterface> injectedInterfaces;
+
+		InjectingClassVisitor(int asmVersion, ClassWriter writer, List<InjectedInterface> injectedInterfaces) {
+			super(asmVersion, writer);
+			this.injectedInterfaces = injectedInterfaces;
+		}
+
+		@Override
+		public void visit(int version, int access, String name, String signature, String superName, String[] interfaces) {
+			Set<String> modifiedInterfaces = new LinkedHashSet<>(interfaces.length + injectedInterfaces.size());
+			Collections.addAll(modifiedInterfaces, interfaces);
+
+			for (InjectedInterface injectedInterface : injectedInterfaces) {
+				modifiedInterfaces.add(injectedInterface.ifaceName());
+			}
+
+			super.visit(version, access, name, signature, superName, modifiedInterfaces.toArray(new String[0]));
```

Comment:
You are not fixing the signature here, if the signature is null, u can ignore it more or less, but if it's not then u will need to fix the signature, this can be done just by appending the injected interface's descriptor to the end of the signature

Reply:
Alright, I tried my best at amending the signature. Does this seem sane to you?